File: src/FlowControl.hs
-- for loops in haskell

import Control.Applicative

-- main1 print 1 to 9
-- in general, we can replace for loop with map
main1 = do
  mapM_ (\x -> print x) [1..9]

nTimes :: Int -> IO () -> IO ()
nTimes 0 do_this = return ()
nTimes n do_this = do
    do_this
    nTimes (n-1) do_this

main2 = do
  mapM_ prettyPrint [1..30]
  where prettyPrint x = putStrLn $ (show x) ++ " ---- " ++ (show $ investmentReturn 10000 0.01 x) 

investmentReturn principle rate day = principle * (1+rate)^day

main3 = do
  x <- (\n -> read n::Int) <$> getLine
  dowhile x

-- similar to nTimes
dowhile x = do
  print x
  if x >= 40
  then return ()
  else dowhile (x + 1)

getNumber = getLine >>= return . (\x->read x::Int)

-- switch or case
main = do
  x <- getNumber
  case x of
    16 -> putStrLn "buy beer"
    21 -> putStrLn "lottery!!"
    _ -> print "default"

File: src/FlowControl.cpp
#include <iostream>
#include <cmath> // need power function
using namespace std;

// main1 print 1 to 9
int main1(){
  
  // iniitialization; loop end; increment
  for (int x = 1; x < 10; x++){
    cout << x << endl;
  }
  return 0;

}

// interest rate calculation
int main2(){

  float amount;
  float principle = 10000;
  float rate = 0.05;   

  for (int day = 1; day <= 30; day++){
    amount = principle * pow(1+rate, day);
    cout << day << " --- " << amount << endl;
  }
  return 0;
}

// do while loop; the purpose of having do-while loop
// the first number guarantee to run.
int main3(){
  int x;
  
  cin >> x;
  do{
  cout << x << endl;
  x++;
  }while(x<50);
  return 0;
}

// switch statement; replace if else statement
// BTW, does R have a switch?
int main(){
  int age;

  cin >> age;
  
  switch(age){
    case 16: 
      cout << "print 16" << endl;
      break; // if no break, will run default
    case 18:
      cout << "go buy lottery tickets" << endl;
      break;
    case 21:
      cout << "buy beer" << endl;
      break;
    default: 
      cout << "this is a default" << endl;
      // end of switch without break
  }
  return 0;
}
